File: provisioning/src/utils/tool_registry.py
"""
Dynamic tool discovery and registration system for MCP VoipNow Provisioning.

This module automatically discovers and registers tools from the tools directory,
eliminating the need for manual imports and registrations in main.py.
"""

import sys
import importlib
import pkgutil
from typing import Dict, List, Callable, Awaitable
import mcp.types as types
import logging
import re

# Type alias for tool handlers
ToolHandler = Callable[[dict, dict, logging.Logger], Awaitable[List[types.TextContent | types.ImageContent | types.EmbeddedResource]]]


class DynamicToolRegistry:
    """
    Dynamically discovers and registers tools from the tools directory.
    
    This class scans the tools directory structure and automatically imports
    modules that contain tool definitions, extracting their schemas and handlers.
    """
    
    def __init__(self, tools_package_path: str = "tools"):
        """
        Initialize the dynamic tool registry.
        
        Args:
            tools_package_path: The package path to the tools directory (e.g., "tools")
        """
        self.tools_package_path = tools_package_path
        self.tool_schemas: List[types.Tool] = []
        self.tool_handlers: Dict[str, ToolHandler] = {}
        self._discovered_modules = set()
        self.logger = logging.getLogger(__name__)
        
    def discover_and_register_tools(self) -> None:
        """
        Discover and register all tools from the tools directory.
        
        This method:
        1. Walks through the tools directory structure
        2. Imports each Python module
        3. Extracts TOOL_REGISTRY configurations
        4. Builds tool schemas and handlers
        """
        self.tool_schemas.clear()
        self.tool_handlers.clear()
        self._discovered_modules.clear()
        
        # Get the tools package
        try:
            tools_package = importlib.import_module(self.tools_package_path)
        except ImportError as e:
            raise ImportError(f"Could not import tools package '{self.tools_package_path}': {e}")
        
        # Walk through all modules in the tools package
        self._walk_package(tools_package)
        
        self.logger.debug(f"Dynamically discovered {len(self.tool_schemas)} tools from {len(self._discovered_modules)} modules")
    
    def _walk_package(self, package) -> None:
        """
        Recursively walk through a package and discover tools.
        
        Args:
            package: The package to walk through
        """
        package_path = package.__path__
        package_name = package.__name__
        
        # Walk through all modules in the package
        for importer, modname, ispkg in pkgutil.iter_modules(package_path, package_name + "."):
            # Validate module name pattern
            if not re.match(r'^tools\.[a-z_]+(?:\.[a-z_]+)*$', modname):
                self.logger.warning(f"Skipping suspicious module name: {modname}")
                continue
            try:
                # Import the module
                module = importlib.import_module(modname)
                self._discovered_modules.add(modname)
                
                # Check if it's a subpackage and recurse
                if ispkg:
                    self._walk_package(module)
                else:
                    # Process the module for tools
                    self._process_module(module, modname)
                    
            except Exception as e:
                self.logger.debug(f"Could not import module '{modname}': {e}")
                continue
    
    def _process_module(self, module, module_name: str) -> None:
        """
        Process a module to extract tool definitions.
        
        Args:
            module: The imported module
            module_name: The name of the module
        """
        # Check if module has TOOL_REGISTRY
        if not hasattr(module, 'TOOL_REGISTRY'):
            return
            
        tool_registry = getattr(module, 'TOOL_REGISTRY')
        if not isinstance(tool_registry, dict):
            return
            
        self.logger.debug(f"Processing tools from module: {module_name}")
        
        # Process each tool in the registry
        for tool_func_name, tool_config in tool_registry.items():
            try:
                # Create tool schema
                schema = self._create_tool_schema(tool_config)
                self.tool_schemas.append(schema)
                
                # Find and register the handler function
                handler_func = getattr(module, tool_func_name, None)
                if handler_func and callable(handler_func):
                    tool_name = tool_config.get("tool_name", tool_func_name)
                    self.tool_handlers[tool_name] = handler_func
                    self.logger.debug(f"Registered tool: {tool_name} (handler: {tool_func_name})")
                else:
                    self.logger.debug(f"Handler function '{tool_func_name}' not found in {module_name}")
                    
            except Exception as e:
                self.logger.debug(f"Error processing tool '{tool_func_name}' in {module_name}: {e}")
                continue
    
    def _create_tool_schema(self, tool_config: dict) -> types.Tool:
        """
        Create a tool schema from tool configuration.
        
        Args:
            tool_config: The tool configuration dictionary
            
        Returns:
            types.Tool: The MCP tool schema
        """
        return types.Tool(
            name=tool_config["tool_name"],
            description=tool_config["tool_description"],
            inputSchema=tool_config["input_schema"]
        )
    
    def get_tool_schemas(self) -> List[types.Tool]:
        """
        Get all discovered tool schemas.
        
        Returns:
            List[types.Tool]: List of all tool schemas
        """
        return self.tool_schemas.copy()
    
    def get_tool_handlers(self) -> Dict[str, ToolHandler]:
        """
        Get all discovered tool handlers.
        
        Returns:
            Dict[str, ToolHandler]: Dictionary mapping tool names to their handlers
        """
        return self.tool_handlers.copy()
    
    def get_tool_handler(self, tool_name: str) -> ToolHandler:
        """
        Get a specific tool handler by name.
        
        Args:
            tool_name: The name of the tool
            
        Returns:
            ToolHandler: The tool handler function
            
        Raises:
            KeyError: If the tool is not found
        """
        if tool_name not in self.tool_handlers:
            raise KeyError(f"Tool '{tool_name}' not found in registry")
        return self.tool_handlers[tool_name]
    
    def refresh(self) -> None:
        """
        Refresh the tool registry by re-discovering all tools.
        
        This is useful when tools are added, removed, or modified at runtime.
        """
        # Clear the module cache for tools package to ensure fresh imports
        modules_to_remove = [name for name in sys.modules.keys() 
                           if name.startswith(self.tools_package_path)]
        for module_name in modules_to_remove:
            if module_name != self.tools_package_path:  # Keep the root package
                del sys.modules[module_name]
        
        # Re-discover tools
        self.discover_and_register_tools()


# Global instance for easy access
_global_registry = None


def get_registry(tools_package_path: str = "tools") -> DynamicToolRegistry:
    """
    Get the global tool registry instance.
    
    Args:
        tools_package_path: The package path to the tools directory
        
    Returns:
        DynamicToolRegistry: The global registry instance
    """
    global _global_registry
    if _global_registry is None:
        _global_registry = DynamicToolRegistry(tools_package_path)
        _global_registry.discover_and_register_tools()
    return _global_registry


def refresh_registry() -> None:
    """
    Refresh the global tool registry.
    """
    global _global_registry
    if _global_registry is not None:
        _global_registry.refresh()


def get_tool_schemas() -> List[types.Tool]:
    """
    Get all tool schemas from the global registry.
    
    Returns:
        List[types.Tool]: List of all tool schemas
    """
    return get_registry().get_tool_schemas()


def get_tool_handlers() -> Dict[str, ToolHandler]:
    """
    Get all tool handlers from the global registry.
    
    Returns:
        Dict[str, ToolHandler]: Dictionary mapping tool names to their handlers
    """
    return get_registry().get_tool_handlers()


def get_tool_handler(tool_name: str) -> ToolHandler:
    """
    Get a specific tool handler by name from the global registry.
    
    Args:
        tool_name: The name of the tool
        
    Returns:
        ToolHandler: The tool handler function
        
    Raises:
        KeyError: If the tool is not found
    """
    return get_registry().get_tool_handler(tool_name)

File: calls/src/tool-registry.ts
/**
 * Dynamic tool discovery and registration system for MCP VoipNow Calls.
 * 
 * This module automatically discovers and registers tools from the tools directory,
 * eliminating the need for manual imports and registrations in index.ts.
 */

import { Tool } from "@modelcontextprotocol/sdk/types.js";
import { Logger } from "winston";
import { readdir, stat } from 'fs/promises';
import fs from 'fs';
import { join, dirname, extname } from "path";
import { fileURLToPath, pathToFileURL } from "url";
import path from 'path';
import { ZodSchema } from "zod";

// Constants for file extensions and naming patterns
const FILE_EXTENSIONS = {
  TYPESCRIPT: '.ts',
  JAVASCRIPT: '.js',
  DECLARATION_TS: '.d.ts',
  DECLARATION_JS: '.d.js',
} as const;

const TOOL_NAMING_PATTERNS = {
  TOOL_SUFFIX: '_TOOL',
  NAME_SUFFIX: 'ToolName',
  DESCRIPTION_SUFFIX: 'ToolDescription',  
  RUNNER_PREFIX: 'run',
  RUNNER_SUFFIX: 'Tool',
  SCHEMA_SUFFIX: 'ToolSchema',
} as const;

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Type definitions for tool modules
export interface ToolModule {
  [key: string]: Tool | string | ToolHandler | ZodSchema<any>;
}

export interface ToolDefinition {
  name: string;
  description: string;
  tool: Tool;
  runner: ToolHandler;
  schema: ZodSchema<any>;
}

export interface ToolHandler {
  (args: any, userAgent: string, config: { voipnowUrl: string; voipnowToken: string; }, logger: Logger): Promise<any>;
}

/**
 * Dynamic tool registry that discovers and manages tools automatically.
 */
export class DynamicToolRegistry {
  private toolDefinitions: ToolDefinition[] = [];
  private toolHandlers: Map<string, ToolHandler> = new Map();
  private logger: Logger;
  private toolsDirectory: string;

  constructor(toolsDirectory: string = "tools", logger?: Logger) {
    // Sanitize and validate the tools directory
    const normalizedDir = path.normalize(toolsDirectory);
    if (normalizedDir.includes('..') || path.isAbsolute(normalizedDir)) {
        throw new Error('Invalid tools directory: must be a relative path without parent directory references');
    }

    const candidatePath = path.join(__dirname, normalizedDir);

    // Verify the directory exists and resolve symlinks
    try {
      const realToolsPath = fs.realpathSync(candidatePath);
      const expectedBase = fs.realpathSync(__dirname);

      // Ensure the real path is within the application directory
      if (!realToolsPath.startsWith(expectedBase + path.sep) && realToolsPath !== expectedBase) {
        throw new Error('Tools directory must be within the application directory');
      }

      // Store the real path for subsequent checks
      this.toolsDirectory = realToolsPath;
    } catch (error: any) {
      throw new Error(`Failed to validate tools directory: ${error.message}`);
    }

    this.logger = logger || console as any;
  }

  /**
   * Discover and register all tools from the tools directory.
   */
  async discoverAndRegisterTools(): Promise<void> {
    this.toolDefinitions = [];
    this.toolHandlers.clear();

    try {
      const toolFiles = await this.findToolFiles(this.toolsDirectory);
      this.logger.debug?.(`Found ${toolFiles.length} tool files`);

      for (const toolFile of toolFiles) {
        try {
          await this.processToolFile(toolFile);
        } catch (error) {
          this.logger.debug?.(`Error processing tool file ${toolFile}: ${error}`);
        }
      }

      this.logger.debug?.(`Dynamically discovered ${this.toolDefinitions.length} tools`);
    } catch (error) {
      throw new Error(`Failed to discover tools in directory '${this.toolsDirectory}': ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Find all TypeScript tool files in the tools directory.
   */
  private async findToolFiles(directory: string): Promise<string[]> {
    const toolFiles: string[] = [];

    try {
      const entries = await readdir(directory);

      for (const entry of entries) {
        const fullPath = join(directory, entry);

        // Verify the path is still within bounds using realpath
        try {
          const realFullPath = fs.realpathSync(fullPath);
          const realToolsDir = fs.realpathSync(this.toolsDirectory);

          // Ensure the file is within the tools directory (prevent symlink escapes)
          if (!realFullPath.startsWith(realToolsDir + path.sep) && realFullPath !== realToolsDir) {
            this.logger.warn?.(`Skipping file outside tools directory: ${entry}`);
            continue;
          }

          const statInfo = await stat(fullPath);

          if (statInfo.isDirectory()) {
            // Recursively search subdirectories
            const subDirectoryFiles = await this.findToolFiles(fullPath);
            toolFiles.push(...subDirectoryFiles);
          } else if (statInfo.isFile()) {
            const ext = extname(entry);
            // Look for .ts files in development or .js files in production
            if ((ext === FILE_EXTENSIONS.TYPESCRIPT && !entry.includes(FILE_EXTENSIONS.DECLARATION_TS)) ||
                (ext === FILE_EXTENSIONS.JAVASCRIPT && !entry.includes(FILE_EXTENSIONS.DECLARATION_JS))) {
              toolFiles.push(fullPath);
            }
          }
        } catch (error) {
          this.logger.debug?.(`Error validating path ${fullPath}: ${error}`);
          continue;
        }
      }
    } catch (error) {
      this.logger.debug?.(`Error reading directory ${directory}: ${error}`);
    }

    return toolFiles;
  }

  /**
   * Process a single tool file and extract tool definitions.
   */
  private async processToolFile(filePath: string): Promise<void> {
    try {
      // Convert file path to file URL for dynamic import
      const fileUrl = pathToFileURL(filePath).href;
      const module: ToolModule = await import(fileUrl);

      // Validate module exports
      if (!this.isValidToolModule(module)) {
        throw new Error(`Invalid tool module structure in ${filePath}`);
      }

      this.logger.debug?.(`Processing tool file: ${filePath}`);

      // Find tool definitions in the module
      const toolDefs = this.extractToolDefinitions(module, filePath);

      for (const toolDef of toolDefs) {
        this.toolDefinitions.push(toolDef);
        this.toolHandlers.set(toolDef.name, toolDef.runner);
        this.logger.debug?.(`Registered tool: ${toolDef.name}`);
      }

    } catch (error) {
      this.logger.debug?.(`Failed to import tool file ${filePath}: ${error}`);
    }
  }

  /**
   * Extract tool definitions from a module by looking for expected patterns.
   */
  private extractToolDefinitions(module: ToolModule, filePath: string): ToolDefinition[] {
    const toolDefs: ToolDefinition[] = [];

    // Look for tool definitions following the naming patterns
    const toolObjects = Object.keys(module).filter(key => 
      key.endsWith(TOOL_NAMING_PATTERNS.TOOL_SUFFIX) && 
      typeof module[key] === 'object' && 
      module[key] !== null &&
      'name' in module[key] &&
      typeof (module[key] as any).name === 'string'
    );

    for (const toolObjectKey of toolObjects) {
      try {
        const toolCandidate = module[toolObjectKey];
        
        // Type guard to ensure this is a Tool object
        if (!this.isValidTool(toolCandidate)) {
          this.logger.debug?.(`Invalid tool object for ${toolObjectKey} in ${filePath}`);
          continue;
        }
        
        const tool = toolCandidate as Tool;
        const toolName = tool.name;

        // Find corresponding name, description, runner, and schema using naming patterns
        const nameKey = Object.keys(module).find(key => 
          key.endsWith(TOOL_NAMING_PATTERNS.NAME_SUFFIX) && 
          typeof module[key] === 'string' &&
          module[key] === toolName
        );
        
        const descriptionKey = Object.keys(module).find(key => 
          key.endsWith(TOOL_NAMING_PATTERNS.DESCRIPTION_SUFFIX) && 
          typeof module[key] === 'string'
        );

        const runnerKey = Object.keys(module).find(key => 
          key.startsWith(TOOL_NAMING_PATTERNS.RUNNER_PREFIX) && 
          key.endsWith(TOOL_NAMING_PATTERNS.RUNNER_SUFFIX) && 
          typeof module[key] === 'function'
        );

        const schemaKey = Object.keys(module).find(key => 
          key.endsWith(TOOL_NAMING_PATTERNS.SCHEMA_SUFFIX) && 
          typeof module[key] === 'object' &&
          module[key] !== null
        );

        if (nameKey && descriptionKey && runnerKey && schemaKey) {
          const name = module[nameKey] as string;
          const description = module[descriptionKey] as string;
          const runner = module[runnerKey] as ToolHandler;
          const schema = module[schemaKey] as ZodSchema<any>;

          const toolDef: ToolDefinition = {
            name,
            description,
            tool,
            runner,
            schema
          };

          toolDefs.push(toolDef);
          this.logger.debug?.(`Extracted tool: ${toolDef.name} from ${filePath}`);
        } else {
          this.logger.debug?.(`Incomplete tool definition for ${toolName} in ${filePath}`);
        }

      } catch (error) {
        this.logger.debug?.(`Error extracting tool from ${toolObjectKey}: ${error}`);
      }
    }

    return toolDefs;
  }

  /**
   * Type guard to validate if an object is a valid Tool.
   */
  private isValidTool(obj: any): obj is Tool {
    return (
      typeof obj === 'object' &&
      obj !== null &&
      typeof obj.name === 'string' &&
      typeof obj.inputSchema === 'object' &&
      obj.inputSchema !== null &&
      typeof obj.inputSchema.type === 'string' &&
      obj.inputSchema.type === 'object'
    );
  }

  /**
   * Validate if a module has valid structure and exports.
   */
  private isValidToolModule(module: any): boolean {
    // Check for required exports and structure
    const hasValidExports = typeof module === 'object' && module !== null;
    const hasNoMaliciousPatterns = !this.containsMaliciousPatterns(module);
    return hasValidExports && hasNoMaliciousPatterns;
  }

  /**
   * Check for potentially malicious patterns in module exports.
   */
  private containsMaliciousPatterns(module: any): boolean {
    const moduleString = JSON.stringify(module);
    const maliciousPatterns = ['eval(', 'Function(', 'require("child_process")', 'exec('];
    return maliciousPatterns.some(pattern => moduleString.includes(pattern));
  }

  /**
   * Get all discovered tool schemas.
   */
  getToolSchemas(): Tool[] {
    return this.toolDefinitions.map(def => def.tool);
  }

  /**
   * Get all discovered tool handlers.
   */
  getToolHandlers(): Map<string, ToolHandler> {
    return new Map(this.toolHandlers);
  }

  /**
   * Get a specific tool handler by name.
   */
  getToolHandler(toolName: string): ToolHandler {
    const handler = this.toolHandlers.get(toolName);
    if (!handler) {
      throw new Error(`Tool '${toolName}' not found in registry`);
    }
    return handler;
  }

  /**
   * Get all tool definitions.
   */
  getToolDefinitions(): ToolDefinition[] {
    return [...this.toolDefinitions];
  }

  /**
   * Refresh the tool registry by re-discovering all tools.
   */
  async refresh(): Promise<void> {
    await this.discoverAndRegisterTools();
  }

  /**
   * Get the number of registered tools.
   */
  getToolCount(): number {
    return this.toolDefinitions.length;
  }
}

// Global registry instance
let globalRegistry: DynamicToolRegistry | null = null;

/**
 * Get the global tool registry instance.
 */
export function getRegistry(toolsDirectory: string = "tools", logger?: Logger): DynamicToolRegistry {
  if (!globalRegistry) {
    globalRegistry = new DynamicToolRegistry(toolsDirectory, logger);
  }
  return globalRegistry;
}

/**
 * Initialize and discover tools using the global registry.
 */
export async function initializeTools(toolsDirectory: string = "tools", logger?: Logger): Promise<DynamicToolRegistry> {
  const registry = getRegistry(toolsDirectory, logger);
  await registry.discoverAndRegisterTools();
  return registry;
}

/**
 * Refresh the global tool registry.
 */
export async function refreshRegistry(): Promise<void> {
  if (globalRegistry) {
    await globalRegistry.refresh();
  }
}

/**
 * Get all tool schemas from the global registry.
 */
export function getToolSchemas(): Tool[] {
  if (!globalRegistry) {
    throw new Error("Tool registry not initialized. Call initializeTools() first.");
  }
  return globalRegistry.getToolSchemas();
}

/**
 * Get all tool handlers from the global registry.
 */
export function getToolHandlers(): Map<string, ToolHandler> {
  if (!globalRegistry) {
    throw new Error("Tool registry not initialized. Call initializeTools() first.");
  }
  return globalRegistry.getToolHandlers();
}

/**
 * Get a specific tool handler by name from the global registry.
 */
export function getToolHandler(toolName: string): ToolHandler {
  if (!globalRegistry) {
    throw new Error("Tool registry not initialized. Call initializeTools() first.");
  }
  return globalRegistry.getToolHandler(toolName);
}
